pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,844 @@
+			h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+			return true
+		}
+
+		// finish
+		if _, err := h.setConfiguredCondition(settingCopy, true, ReasonCompleted, ""); err != nil {
+			logrus.Warnf("update status error %v", err)
+		}
+		return true
+	}
+	return false
+}
+
+// true: all detach
+func (h *Handler) checkLonghornVolumeDetach() (bool, error) {
+	volumes, err := h.longhornVolumeCache.List(util.LonghornSystemNamespaceName, labels.Everything())
+	if err != nil {
+		logrus.Warnf("volume error %v", err)
+		return false, err
+	}
+
+	for _, volume := range volumes {
+		if volume.Status.State != "detached" {
+			logrus.Infof("volume state: %v", volume.Status.State)
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (h *Handler) checkPrometheusStatusAndStart() bool {
+	// check prometheus cattle-monitoring-system/rancher-monitoring-prometheus replica
+	prometheus, err := h.prometheusCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringPrometheus)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("prometheus not found. skip")
+			return true
+		}
+		logrus.Warnf("prometheus get error %v", err)
+		return false
+	}
+
+	// check started or not
+	if replicasStr, ok := prometheus.Annotations[ReplicaStorageNetworkAnnotation]; ok {
+		logrus.Infof("current prometheus replicas: %v", *prometheus.Spec.Replicas)
+		logrus.Infof("start prometheus")
+		prometheusCopy := prometheus.DeepCopy()
+		// ignore error, and keep original replica
+		if replicas, err := strconv.Atoi(replicasStr); err == nil {
+			*prometheusCopy.Spec.Replicas = int32(replicas)
+		}
+		delete(prometheusCopy.Annotations, ReplicaStorageNetworkAnnotation)
+
+		if _, err := h.prometheus.Update(prometheusCopy); err != nil {
+			logrus.Warnf("prometheus update error %v", err)
+			return false
+		}
+		return true
+	}
+
+	return true
+}
+
+func (h *Handler) checkAltermanagerStatusAndStart() bool {
+	// check alertmanager cattle-monitoring-system/rancher-monitoring-alertmanager replica
+	alertmanager, err := h.alertmanagerCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringAlertmanager)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("Alertmanager not found. skip")
+			return true
+		}
+		logrus.Warnf("alertmanager get error %v", err)
+		return false
+	}
+
+	// check started or not
+	if replicasStr, ok := alertmanager.Annotations[ReplicaStorageNetworkAnnotation]; ok {
+		logrus.Infof("current alertmanager replicas: %v", *alertmanager.Spec.Replicas)
+		logrus.Infof("start alertmanager")
+		alertmanagerCopy := alertmanager.DeepCopy()
+		// ignore error, and keep original replica
+		if replicas, err := strconv.Atoi(replicasStr); err == nil {
+			*alertmanagerCopy.Spec.Replicas = int32(replicas)
+		}
+		delete(alertmanagerCopy.Annotations, ReplicaStorageNetworkAnnotation)
+
+		if _, err := h.alertmanager.Update(alertmanagerCopy); err != nil {
+			logrus.Warnf("alertmanager update error %v", err)
+			return false
+		}
```

Comment:
In k8s controller, we should leverage its reconciliation mechanism and return the error if the output is not expected, e.g., return the error directly if the update is failed and the controller will retry automatically, therefore you will get the expected outcome at the final stage.

Reply:
The problem is 
if we return error and let onChange function to reconcile again.
We have to early return in `checkPodStatusAndStart` otherwise we cannot get the correct error message for reconcile.
Instead, I logged error in function and return state back to caller to ask reconcile.

Reply:
If we saved the error and continue to run next function, but there are multiple errors in checkPodStatusAndStart,
Which one should we return back to the caller?

Reply:
if the first one is failed it should stop and return immediately, the main concept of controller onChange is that is should be idempotent and reconcilable.